File: standalone/distance/distance-tf.py
#!/usr/bin/env python

import numpy as np
import math
import tensorflow as tf

def create_data( count = 10 ):
    inputs = np.random.random([count,3]);
    inputs.sort()
    return inputs, np.array([ [math.sqrt((v*v).sum())] for v in inputs ])

def distance_tf():
    # the hyperparameters

    learning_rate = .0001 
    batch_size    = 3
    epochs        = 10

    training_size = 1000 * 17
    testing_size  = 1000 * 3

    # create the model

    model = tf.keras.Sequential() # not tf.keras.models.Sequential() apparently...
    model.add( tf.keras.layers.Dense(units=1, input_shape=(3,)))

    model.compile(
        optimizer = tf.keras.optimizers.RMSprop( learning_rate = learning_rate )
        , loss    = "mean_squared_error"
        , metrics = [ tf.keras.metrics.RootMeanSquaredError() ]
    )

    # train the model

    training_inputs, training_labels = create_data( training_size )

    history = model.fit(
          x = training_inputs
        , y = training_labels
        , batch_size = batch_size
        , epochs = epochs
        , shuffle = True
    )
    #print( 'history', history )
    #print( 'history.history', history.history )
    #print( 'history.epoch', history.epoch )
    for weight in model.weights:
        print( '>weight', np.array( weight ).flatten() )

    # test the model

    testing_inputs, testing_labels = create_data( testing_size )

    model.evaluate( 
          x = testing_inputs
        , y = testing_labels
        , batch_size = batch_size
    )

    # the end

distance_tf()

# eof
##############################################################################

File: standalone/distance/distance-tf.js
#!/usr/bin/env node 

const tf = require('@tensorflow/tfjs-node-gpu');

const create_data = ( count = 10 ) => {
	const labels = new Array( count );
	const data = { 
		labels:labels,
		inputs: new Array( count ).fill( 0 ).map( (_,i) => {
			const v = new Array( 3 ).fill( 0 ).map( _=> Math.random() ).sort( (a,b) => a-b );
			labels[ i ] = [Math.sqrt( v.reduce( (s,x)=>s+x*x, .0 ) )];
			return v;
		})
	}
	return data;
};

const distance_tf = async () => {
	// the hyperparameters

	const learningRate = .0001;
	const batchSize	= 3;
	const epochs	   = 10;

	const trainingSize = 1000 * 17;
	const testingSize  = 1000 * 3;

	// create the model

	const model = tf.sequential();
	model.add( tf.layers.dense( { units:1, inputShape:[3,] } ) )

	model.compile({
		  optimizer: tf.train.rmsprop( learningRate )
		, loss: 'meanSquaredError'
		, metrics: [ tf.metrics.meanSquaredError ]
	})

	// train the model

	const training = create_data( trainingSize )

	const history = await model.fit(
		  tf.tensor2d( training.inputs )
		, tf.tensor2d( training.labels )
		, {
			  batchSize : batchSize
			, epochs : epochs
			, shuffle : true
		}
	);

	//console.log( 'history', history );
	//console.log( 'history.history', history.history );
	//console.log( 'history.epoch', history.epoch );
	model.weights.forEach( weight => {
		console.log( '>weight', weight.read().arraySync().flat() );
	});

	// test the model

	testing = create_data( testingSize );

	await model.evaluate( 
		  tf.tensor2d( testing.inputs )
		, tf.tensor2d( testing.labels )
		, batchSize 
	);

	// the end
};

distance_tf()

// eof
//////////////////////////////////////////////////////////////////////////////
